include/nbl/video/IDescriptorPool.h

```diff
@@ -32,86 +30,186 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+                baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_accelerationStructureStorage.get());
+                break;
+            default:
+                assert(!"Invalid code path.");
+                return nullptr;
+            }
+
+            return baseAddress;
+        }
+
+        inline core::smart_refctd_ptr<IGPUSampler>* getMutableSamplerStorage() const
         {
-            std::fill_n(m_maxDescriptorCount, asset::EDT_COUNT, 0u);
+            return reinterpret_cast<core::smart_refctd_ptr<IGPUSampler>*>(m_mutableSamplerStorage.get());
+        }
 
-            for (auto i = 0; i < poolSizeCount; ++i)
-                m_maxDescriptorCount[poolSizes[i].type] += poolSizes[i].count;
+        friend class IGPUDescriptorSet;
+        // Returns the offset into the pool's descriptor storage. These offsets will be combined
+        // later with base memory addresses to get the actual memory address where we put the core::smart_refctd_ptr<const IDescriptor>.
+        SDescriptorOffsets allocateDescriptorOffsets(const IGPUDescriptorSetLayout* layout);
 
-            for (auto i = 0; i < asset::EDT_COUNT; ++i)
+        struct allocator_state_t
+        {
+            allocator_state_t(const uint32_t maxDescriptorCount, const bool allowsFreeing)
             {
-                if (m_maxDescriptorCount[i] > 0)
+                if (maxDescriptorCount == 0)
+                    return;
+
+                if (allowsFreeing)
                 {
-                    if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
-                    {
-                        m_generalAllocatorReservedSpace[i] = std::make_unique<uint8_t[]>(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(1u, m_maxDescriptorCount[i], 1u));
-                        m_generalAllocators[i] = core::GeneralpurposeAddressAllocator<uint32_t>(m_generalAllocatorReservedSpace[i].get(), 0u, 0u, 1u, m_maxDescriptorCount[i], 1u);
-                    }
-                    else
-                    {
-                        m_linearAllocators[i] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[i]);
-                    }
+                    generalAllocatorReservedSpace = std::make_unique<uint8_t[]>(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(1u, maxDescriptorCount, 1u));
+                    generalAllocator = core::GeneralpurposeAddressAllocator<uint32_t>(generalAllocatorReservedSpace.get(), 0u, 0u, 1u, maxDescriptorCount, 1u);
+                }
+                else
+                {
+                    linearAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, maxDescriptorCount);
                 }
             }
 
-            // For (possibly) mutable samplers.
-            if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+            ~allocator_state_t() {}
+
+            inline uint32_t allocate(const uint32_t count, const bool allowsFreeing)
```

Comment:
you don't need `allowsFreeing` you can deduct that from `generalAllocatorReservedSpace` not being a nullptr

Reply:
Resolved.